README
NAME
    PPI::PowerToys - A handy collection of small PPI-based utilities

DESCRIPTION
    The PPI PowerToys are a small collection of utilities for working with
    Perl files, modules and distributions.

    To kick off the collection, he's added a very simple and raw version of
    one of his own little tools.

  ppi_version
      > ppi_version show
      > ppi_version change 0.01 0.02

    ppi_version is a utility for working with version numbers in groups of
    modules.

    The "ppi_version show" command will scan through your distribution
    (starting in the current directory and working down) and locate all the
    versions for the various Perl files in the distribution.

    It scans through all files inside the current directory with one-only
    instance of the line.

      $VERSION = '0.01';

    The "ppi_version change" command scans through your distribution
    (starting in the current directory and working down) and locate all
    cases where the $VERSION is the first param, replacing it (safely) with
    the second params.

TO DO
    - Add extra commands to ppi_version

    - Include more useful utilities

    - Any improvements to the current utilities are also very welcome.

SUPPORT
    Bugs and patches should be reported via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PPI-PowerToys>

    For other issues, contact the author.

AUTHOR
    Adam Kennedy <adamk@cpan.org>

COPYRIGHT
    Copyright 2006 - 2008 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension Perl-PowerToys

0.13 to be released
	- Upgrading to Module::Install::DSL 0.87
	- Adding new ppi_copyright tool

0.12 Thu 30 Apr 2009
	- Upgrading to Module::Install::DSL 0.83
	- Adding support for "our $VERSION";
	- Adding support for all four quote variants

0.11 Thu  6 Mar 2008
	- Upgrading to Module::Install 0.76
	- Adding increment support
	- Bug fixes

0.10 Sun 16 Nov 2007
	- VERSION detection supports fully referenced variables
	  (now in the change command as well)
	- Refactored ppi_version into PPI::App::Version

0.09 Fri  2 Nov 2007
	- VERSION detection supports fully referenced variables

0.08 Thu 12 Jul 2007
        - If PPI::Document fails for a document, skip it
        - Add a Usage message if no command
        - Add a show command

0.07 Wed  9 May 2007
        - Upgrading to use File::Find::Rule::Perl
        - Upgrading to Module::Install 0.67

0.06 Sun 19 Mar 2006
	- Renaming to PPI-PowerToys for better clarity
	- Changing ppichangeversion to ppi_version to allow the addition
	  or more features in future versions.

0.05 Sun 22 Jan 2006
	- Skip files we can't write to

0.04 Sun 15 Jan 2006
	- Allow _ in ppichangeversion versions

0.03 Wed 11 Jan 2006
	- Added .t files to the list checked

0.01 Wed 11 Jan 2006
	- Created initial version


lib/PPI/App/ppi_copyright.pm
package PPI::App::ppi_copyright;

use 5.006;
use strict;
use warnings;
use version                0.74 ();
use File::Spec             0.80 ();
use PPI::Document         1.201 ();
use File::Find::Rule       0.30 ();
use File::Find::Rule::Perl 0.03 ();

use vars qw{$VERSION};
BEGIN {
        $VERSION = '0.13';
}





#####################################################################
# Main Functions

sub main {
	my $cmd = shift @_;
	return usage(@_)  unless defined $cmd;
	return show(@_)   if $cmd eq 'show';
	return change(@_) if $cmd eq 'change';
	return error("Unknown command '$cmd'");
}

sub error {
	my $msg = shift;
	chomp $msg;
	print "\n";
	print "  $msg\n";
	print "\n";
	return 255;
}





#####################################################################
# Command Functions

sub usage {
	print "\n";
	print "ppi_version $VERSION - Copyright 2006 - 2009 Adam Kennedy.\n";
	print "Usage:\n";
	print "  ppi_version show\n";
	print "  ppi_version change 0.02_03 0.54\n";
	print "\n";
	return 0;
}

sub show {
	# Find all modules and scripts below the current directory
	my @files = File::Find::Rule->perl_file->in( File::Spec->curdir );
	print "Found " . scalar(@files) . " file(s)\n";

	my $count = 0;
	foreach my $file ( @files ) {
		print "$file...";
		my $document = PPI::Document->new($file);
		unless ( $document ) {
			print " failed to parse file\n";
			next;
		}

		# Does the document contain a simple version number
		my $elements = $document->find( \&_wanted );
		unless ( $elements ) {
			print " no copyright\n";
			next;
		}

		if ( @$elements ) {
			# Print the raw copyright lines
			print "\n";
			print "\n";
			foreach my $element ( @$elements ) {
				my $pod = $element->content;
				print map {
					"  $_\n"
				} grep {
					/Copyright/
				} split /\n/, $pod;
			}
			print "\n";
			$count++;
		}
	}

	print "Found " . scalar($count) . " copyright(s)\n";
	print "Done.\n";
	return 0;	
}

sub change {
	# Find all modules and scripts below the current directory
	my @files = File::Find::Rule->perl_file->in( File::Spec->curdir );
	print "Found " . scalar(@files) . " file(s)\n";

	my $count = 0;
	foreach my $file ( @files ) {
		print "$file...";
		if ( ! -w $file ) {
			print " no write permission\n";
			next;
		}
		my $rv = _change_file( $file );
		if ( $rv ) {
			print " updated\n";
			$count++;
		} elsif ( defined $rv ) {
			print " skipped\n";
		} else {
			print " failed to parse file\n";
		}
	}

	print "Updated " . scalar($count) . " file(s)\n";
	print "Done.\n";
	return 0;
}






#####################################################################
# Support Functions

sub _change_file {
	my $file = shift;

	# Parse the file
	my $document = PPI::Document->new($file);
	unless ( $document ) {
		error("Failed to parse $file");
	}

	# Apply the changes
	my $rv = _change_document( $document );
	unless ( defined $rv ) {
		error("$file contains more than one \$VERSION assignment");
	}
	unless ( $rv ) {
		return '';
	}

	# Save the updated version
	unless ( $document->save($file) ) {
		error("PPI::Document save failed");
	}

	return 1;
}

sub _change_document {
	my $document = shift;

	# Does the document contain an element
	my $elements = $document->find( \&_wanted );
	unless ( $elements and @$elements ) {
		return '';
	}

	my $pattern = qr/\b(copyright\s+\d{4}(?:\s*-\s*\d{4}))/i;
	foreach my $element ( @$elements ) {
		$element->{content} =~ s/$pattern/_change($1)/eg;
	}

	return 1;
}

# Locate a version number token
sub _wanted {
	return !! (
		$_[1]->isa('PPI::Token::Pod')
		and
		$_[1]->content =~ /\bCopyright\b/
	);
}

sub _change {
	my $copyright = shift;
	my $thisyear  = (localtime time)[5] + 1900;
	my @year      = $copyright =~ m/(\d{4})/g;

	if ( @year == 1 ) {
		# Handle the single year format
		if ( $year[0] == $thisyear ) {
			# No change
			return $copyright;
		} else {
			# Convert from single year to multiple year
			$copyright =~ s/(\d{4})/$1 - $thisyear/;
			return $copyright;
		}
	}

	if ( @year == 2 ) {
		# Handle the range format
		if ( $year[1] == $thisyear ) {
			# No change
			return $copyright;
		} else {
			# Change the second year to the current one
			$copyright =~ s/$year[1]/$thisyear/;
			return $copyright;
		}
	}

	# huh?
	die "Invalid or unknown copyright line $copyright";
}

1;


inc/Module/Install/Scripts.pm
#line 1
package Module::Install::Scripts;

use strict;
use Module::Install::Base;
use File::Basename ();

use vars qw{$VERSION @ISA};
BEGIN {
	$VERSION = '0.62';
	@ISA     = qw(Module::Install::Base);
}

sub prompt_script {
    my ($self, $script_file) = @_;

    my ($prompt, $abstract, $default);
    foreach my $line ( $self->_read_script($script_file) ) {
        last unless $line =~ /^#/;
        $prompt = $1   if $line =~ /^#\s*prompt:\s+(.*)/;
        $default = $1  if $line =~ /^#\s*default:\s+(.*)/;
        $abstract = $1 if $line =~ /^#\s*abstract:\s+(.*)/;
    }
    unless (defined $prompt) {
        my $script_name = File::Basename::basename($script_file);
        $prompt = "Do you want to install '$script_name'";
        $prompt .= " ($abstract)" if defined $abstract;
        $prompt .= '?';
    }
    return unless $self->prompt($prompt, ($default || 'n')) =~ /^[Yy]/;
    $self->install_script($script_file);
}

sub install_script {
    my $self = shift;
    my $args = $self->makemaker_args;
    my $exe_files = $args->{EXE_FILES} ||= [];
    push @$exe_files, @_;
}

sub _read_script {
    my ($self, $script_file) = @_;
    local *SCRIPT;
    open SCRIPT, $script_file
      or die "Can't open '$script_file' for input: $!\n";
    return <SCRIPT>;
}

1;


lib/PPI/App/ppi_version.pm
package PPI::App::ppi_version;

use 5.006;
use strict;
use warnings;
use version                0.74 ();
use File::Spec             0.80 ();
use PPI::Document         1.201 ();
use File::Find::Rule       0.30 ();
use File::Find::Rule::Perl 0.03 ();

use vars qw{$VERSION};
BEGIN {
        $VERSION = '0.13';
}





#####################################################################
# Main Functions

sub main {
	my $cmd = shift @_;
	return usage(@_)  unless defined $cmd;
	return show(@_)   if $cmd eq 'show';
	return change(@_) if $cmd eq 'change';
	return error("Unknown command '$cmd'");
}

sub error {
	my $msg = shift;
	chomp $msg;
	print "\n";
	print "  $msg\n";
	print "\n";
	return 255;
}





#####################################################################
# Command Functions

sub usage {
	print "\n";
	print "ppi_version $VERSION - Copyright 2006 - 2009 Adam Kennedy.\n";
	print "Usage:\n";
	print "  ppi_version show\n";
	print "  ppi_version change 0.02_03 0.54\n";
	print "\n";
	return 0;
}

sub show {
	# Find all modules and scripts below the current directory
	my @files = File::Find::Rule->perl_file->in( File::Spec->curdir );
	print "Found " . scalar(@files) . " file(s)\n";

	my $count = 0;
	foreach my $file ( @files ) {
		print "$file...";
		my $document = PPI::Document->new($file);
		unless ( $document ) {
			print " failed to parse file\n";
			next;
		}

		# Does the document contain a simple version number
		my $elements = $document->find( \&_wanted );
		unless ( $elements ) {
			print " no version\n";
			next;
		}
		if ( @$elements > 1 ) {
			error("$file contains more than one \$VERSION");
		}

		# What is that number
		my $version = _get_version($elements->[0]);
		unless ( defined $version ) {
			error("Failed to get version string");
		}
		print " $version\n";
		$count++;
	}

	print "Found " . scalar($count) . " version(s)\n";
	print "Done.\n";
	return 0;	
}

sub change {
	my $from = shift @_;
	unless ( $from and $from =~ /^[\d\._]+$/ ) {
		error("From is not a number");
	}
	my $to = shift @_;
	unless ( $to and $to =~ /^[\d\._]+$/ ) {
		error("To is not a number");
	}

	# Find all modules and scripts below the current directory
	my @files = File::Find::Rule->perl_file->in( File::Spec->curdir );
	print "Found " . scalar(@files) . " file(s)\n";

	my $count = 0;
	foreach my $file ( @files ) {
		print "$file...";
		if ( ! -w $file ) {
			print " no write permission\n";
			next;
		}
		my $rv = _change_file( $file, $from => $to );
		if ( $rv ) {
			print " updated\n";
			$count++;
		} elsif ( defined $rv ) {
			print " skipped\n";
		} else {
			print " failed to parse file\n";
		}
	}

	print "Updated " . scalar($count) . " file(s)\n";
	print "Done.\n";
	return 0;
}






#####################################################################
# Support Functions

sub _change_file {
	my $file = shift;
	my $from = shift;
	my $to   = shift;

	# Parse the file
	my $document = PPI::Document->new($file);
	unless ( $document ) {
		error("Failed to parse $file");
	}

	# Apply the changes
	my $rv = _change_document( $document, $from => $to );
	unless ( defined $rv ) {
		error("$file contains more than one \$VERSION assignment");
	}
	unless ( $rv ) {
		return '';
	}

	# Save the updated version
	unless ( $document->save($file) ) {
		error("PPI::Document save failed");
	}

	return 1;
}

sub _change_document {
	my $document = shift;
	my $from     = shift;
	my $to       = shift;

	# Does the document contain an element
	my $elements = $document->find( \&_wanted );
	unless ( $elements ) {
		return '';
	}
	if ( @$elements > 1 ) {
		return undef;
	}

	# Find (and if it matches, replace) the version
	my $version = _get_version($elements->[0]);
	unless ( $version eq $from ) {
		return '';
	}

	# Set the new version
	_set_version( $elements->[0], $to );

	return 1;
}

# Extract the version
sub _get_version {
	my $token = shift;
	if ( $token->isa('PPI::Token::Quote') ) {
		if ( $token->can('literal') ) {
			return $token->literal;
		} else {
			return $token->string;
		}
	} elsif ( $token->isa('PPI::Token::Number') ) {
		if ( $token->can('literal') ) {
			return $token->literal;
		} else {
			return $token->content;
		}
	}
	die('Unsupported object ' . ref($token));
}

# Change the version.
# We need to hack some internals to achieve this,
# but it will have to do for now.
sub _set_version {
	my $token = shift;
	my $to    = shift;
	if ( $token->isa('PPI::Token::Number') ) {
		$token->{content} = $to;
	} elsif ( $token->isa('PPI::Token::Quote::Single') ) {
		$token->{content} = qq|'$to'|;
	} elsif ( $token->isa('PPI::Token::Quote::Double') ) {
		$token->{content} = qq|"$to"|;
	} elsif ( $token->isa('PPI::Token::Quote::Literal') ) {
		substr(
			$token->{content},
			$token->{sections}->[0]->{position},
			$token->{sections}->[0]->{size},
			$to,
		);
	} elsif ( $token->isa('PPI::Token::Quote::Interpolate') ) {
		substr(
			$token->{content},
			$token->{sections}->[0]->{position},
			$token->{sections}->[0]->{size},
			$to,
		);
	} else {
		die('Unsupported object ' . ref($token));
	}
	return 1;
}

sub _file_version {
	my $file = shift;
	my $doc  = PPI::Document->new($file);
	unless ( $doc ) {
		return "failed to parse file";
	}

	# Does the document contain a simple version number
	my $elements = $doc->find( \&_find_version );
	unless ( $elements ) {
		return "no version";
	}
	if ( @$elements > 1 ) {
		error("$file contains more than one \$VERSION");
	}
	my $element = $elements->[0];
	my $version = $element->snext_sibling->snext_sibling;
	my $version_string = $version->string;
	unless ( defined $version_string ) {
		error("Failed to get version string");
	}

	return version->new($version_string);
}

# Locate a version number token
sub _wanted {
	# Must be a quote or number
	$_[1]->isa('PPI::Token::Quote')          or
	$_[1]->isa('PPI::Token::Number')         or return '';

	# To the right is a statement terminator or nothing
	my $t = $_[1]->snext_sibling;
	if ( $t ) {
		$t->isa('PPI::Token::Structure') or return '';
		$t->content eq ';'               or return '';
	}

	# To the left is an equals sign
	my $e = $_[1]->sprevious_sibling         or return '';
	$e->isa('PPI::Token::Operator')          or return '';
	$e->content eq '='                       or return '';

	# To the left is a $VERSION symbol
	my $v = $e->sprevious_sibling            or return '';
	$v->isa('PPI::Token::Symbol')            or return '';
	$v->content =~ m/^\$(?:\w+::)*VERSION$/  or return '';

	# To the left is either nothing or "our"
	my $o = $v->sprevious_sibling;
	if ( $o ) {
		$o->content eq 'our'             or return '';
		$o->sprevious_sibling           and return '';
	}

	return 1;
}

1;
